File: q20_stock_trading/solution-1-java/TradingService.java
import java.util.*;

public class TradingService {

    // Preserve insertion order for showStocks()
    private final LinkedHashMap<String, Integer> stocks = new LinkedHashMap<>();
    private final Map<String, User> users = new HashMap<>();
    // Track OPEN LIMIT orders per stock for instant re-eval on price updates
    private final Map<String, List<OrderRef>> openLimitByStock = new HashMap<>();

    public TradingService() { }

    // Create a user with an initial cash balance
    public void addUser(String userId, int openingCash) {
        users.putIfAbsent(userId, new User(userId, openingCash));
    }

    // Fetch user's current account balance
    public int getAccountBalance(String userId) {
        User u = users.get(userId);
        return (u == null) ? 0 : u.cash;
    }

    // Add/update a stock price; on update, auto-fill favorable OPEN LIMIT orders
    public void addStock(String stockName, int stockPrice) {
        stocks.putIfAbsent(stockName, stockPrice); // preserve first-add order
        stocks.put(stockName, stockPrice);         // update latest price
        autoFillOpenLimits(stockName, stockPrice);
    }

    // Show all stocks as "name price" in first-added order
    public List<String> showStocks() {
        List<String> out = new ArrayList<>();
        for (Map.Entry<String, Integer> e : stocks.entrySet()) {
            out.add(e.getKey() + " " + e.getValue());
        }
        return out;
    }

    // Place an order; returns FILLED, OPEN, or REJECTED
    public String submitOrder(String userId, String stockName, String orderKind, String action, int qty, int limitPrice) {
        User user = users.get(userId);
        Integer currentPrice = stocks.get(stockName);

        // Basic existence/qty check; if missing, reject (record if user exists)
        if (user == null) {
            return Status.REJECTED.name();
        }
        if (currentPrice == null || qty <= 0) {
            OrderKind kind = OrderKind.valueOf(orderKind);
            Action act = Action.valueOf(action);
            Order ord = new Order(stockName, act, qty, kind, Status.REJECTED, limitPrice);
            user.orders.add(ord);
            return ord.status.name();
        }

        OrderKind kind = OrderKind.valueOf(orderKind);
        Action act = Action.valueOf(action);
        int priceAtSubmit = currentPrice;

        if (kind == OrderKind.MARKET) {
            return handleMarket(user, stockName, act, qty, priceAtSubmit);
        } else { // LIMIT
            return handleLimit(user, stockName, act, qty, limitPrice, priceAtSubmit);
        }
    }

    // Show holdings for a user as "stockName count" in the order first acquired (ever)
    public List<String> viewStockHoldings(String userId) {
        User user = users.get(userId);
        if (user == null) return Collections.emptyList();
        List<String> out = new ArrayList<>();
        for (String sym : user.firstAcquired) { // stable “first-ever” acquisition order
            Holding h = user.holdings.get(sym);
            if (h != null && h.qty > 0) {
                out.add(sym + " " + h.qty);
            }
        }
        return out;
    }

    // Show order history as "stockName action quantity orderKind status"
    public List<String> viewOrders(String userId) {
        User user = users.get(userId);
        if (user == null) return Collections.emptyList();
        List<String> out = new ArrayList<>();
        for (Order o : user.orders) {
            out.add(o.stockName + " " + o.action.name() + " " + o.qty + " " + o.kind.name() + " " + o.status.name());
        }
        return out;
    }

    // --- Helpers ------------------------------------------------------------

    private String handleMarket(User user, String stock, Action act, int qty, int priceAtSubmit) {
        Status status;
        if (act == Action.BUY) {
            long cost = 1L * qty * priceAtSubmit;
            if (user.cash >= cost) {
                applyBuy(user, stock, qty, priceAtSubmit);
                status = Status.FILLED;
            } else {
                status = Status.REJECTED;
            }
        } else { // SELL
            int held = user.getHoldingQty(stock);
            if (held >= qty) {
                applySell(user, stock, qty, priceAtSubmit);
                status = Status.FILLED;
            } else {
                status = Status.REJECTED;
            }
        }
        Order ord = new Order(stock, act, qty, OrderKind.MARKET, status, 0);
        user.orders.add(ord);
        return status.name();
    }

    private String handleLimit(User user, String stock, Action act, int qty, int limitPrice, int priceAtSubmit) {
        Status status;

        if (act == Action.BUY) {
            // Validation requires enough cash at CURRENT price even if not favorable
            long costAtCurrent = 1L * qty * priceAtSubmit;
            if (user.cash < costAtCurrent) {
                status = Status.REJECTED;
            } else if (priceAtSubmit <= limitPrice) {
                applyBuy(user, stock, qty, priceAtSubmit);
                status = Status.FILLED;
            } else {
                status = Status.OPEN;
            }
        } else { // SELL
            int held = user.getHoldingQty(stock);
            if (held < qty) {
                status = Status.REJECTED;
            } else if (priceAtSubmit >= limitPrice) {
                applySell(user, stock, qty, priceAtSubmit);
                status = Status.FILLED;
            } else {
                status = Status.OPEN;
            }
        }

        Order ord = new Order(stock, act, qty, OrderKind.LIMIT, status, limitPrice);
        user.orders.add(ord);
        if (status == Status.OPEN) {
            openLimitByStock.computeIfAbsent(stock, k -> new ArrayList<>())
                            .add(new OrderRef(user.userId, ord));
        }
        return status.name();
    }

    // Auto-fill OPEN LIMIT orders for this stock if the new price is favorable
    private void autoFillOpenLimits(String stockName, int newPrice) {
        List<OrderRef> refs = openLimitByStock.get(stockName);
        if (refs == null || refs.isEmpty()) return;

        Iterator<OrderRef> it = refs.iterator();
        while (it.hasNext()) {
            OrderRef ref = it.next();
            User user = users.get(ref.userId);
            Order ord = ref.order;
            if (user == null || ord.status != Status.OPEN || ord.kind != OrderKind.LIMIT) {
                it.remove();
                continue;
            }

            boolean favorable = (ord.action == Action.BUY) ? (newPrice <= ord.limitPrice)
                                                           : (newPrice >= ord.limitPrice);
            if (!favorable) continue;

            if (ord.action == Action.BUY) {
                long cost = 1L * ord.qty * newPrice;
                if (user.cash >= cost) {
                    applyBuy(user, stockName, ord.qty, newPrice);
                    ord.status = Status.FILLED;
                    it.remove();
                }
                // else: keep OPEN (insufficient cash at fill time)
            } else { // SELL
                int held = user.getHoldingQty(stockName);
                if (held >= ord.qty) {
                    applySell(user, stockName, ord.qty, newPrice);
                    ord.status = Status.FILLED;
                    it.remove();
                }
                // else: keep OPEN (insufficient qty at fill time)
            }
        }
        if (refs.isEmpty()) openLimitByStock.remove(stockName);
    }

    private void applyBuy(User user, String stock, int qty, int price) {
        long cost = 1L * qty * price;
        user.cash -= (int) cost;

        Holding h = user.holdings.computeIfAbsent(stock, s -> new Holding());
        // Record first-ever acquisition order (stable, even if later sold to zero)
        if (!user.firstAcquired.contains(stock)) {
            user.firstAcquired.add(stock);
        }

        long totalCostBefore = 1L * h.avgPrice * h.qty;
        int newQty = h.qty + qty;
        int newAvg = (int) ((totalCostBefore + cost) / newQty);
        h.qty = newQty;
        h.avgPrice = newAvg;
    }

    private void applySell(User user, String stock, int qty, int price) {
        long proceeds = 1L * qty * price;
        user.cash += (int) proceeds;
        Holding h = user.holdings.get(stock);
        if (h == null || h.qty < qty) return; // should not happen after validation
        h.qty -= qty;
        if (h.qty == 0) {
            user.holdings.remove(stock);
            // do NOT remove from firstAcquired: order must remain “first-ever”
        }
        // avgPrice remains for remaining shares
    }

    // --- Data types ---------------------------------------------------------

    private enum OrderKind { MARKET, LIMIT }
    private enum Action { BUY, SELL }
    private enum Status { FILLED, OPEN, REJECTED }

    private static final class Holding {
        int qty = 0;
        int avgPrice = 0; // maintained for completeness
    }

    private static final class Order {
        final String stockName;
        final Action action;
        final int qty;
        final OrderKind kind;
        Status status;
        final int limitPrice; // for OPEN limits; 0 for MARKET

        Order(String stockName, Action action, int qty, OrderKind kind, Status status, int limitPrice) {
            this.stockName = stockName;
            this.action = action;
            this.qty = qty;
            this.kind = kind;
            this.status = status;
            this.limitPrice = limitPrice;
        }
    }

    private static final class User {
        final String userId;
        int cash;

        // Current holdings by symbol (contents change as user trades)
        final Map<String, Holding> holdings = new HashMap<>();

        // Stable “first-ever acquisition” order per symbol for this user
        final LinkedHashSet<String> firstAcquired = new LinkedHashSet<>();

        // Orders in placement order
        final List<Order> orders = new ArrayList<>();

        User(String userId, int openingCash) {
            this.userId = userId;
            this.cash = openingCash;
        }

        int getHoldingQty(String stock) {
            Holding h = holdings.get(stock);
            return (h == null) ? 0 : h.qty;
        }
    }

    private static final class OrderRef {
        final String userId;
        final Order order;

        OrderRef(String userId, Order order) {
            this.userId = userId;
            this.order = order;
        }
    }
}

File: q20_stock_trading/solution-2-python/TradingService.py
from __future__ import annotations
from dataclasses import dataclass
from enum import Enum
from typing import Dict, List, Optional
from collections import OrderedDict


class OrderKind(Enum):
    MARKET = "MARKET"
    LIMIT = "LIMIT"


class Action(Enum):
    BUY = "BUY"
    SELL = "SELL"


class Status(Enum):
    FILLED = "FILLED"
    OPEN = "OPEN"
    REJECTED = "REJECTED"


@dataclass
class Holding:
    qty: int = 0
    avgPrice: int = 0  # maintained for completeness


@dataclass
class Order:
    stockName: str
    action: Action
    qty: int
    kind: OrderKind
    status: Status
    limitPrice: int  # 0 for MARKET


@dataclass
class OrderRef:
    userId: str
    order: Order


class User:
    def __init__(self, userId: str, openingCash: int) -> None:
        self.userId = userId
        self.cash: int = openingCash
        # Current holdings by symbol (contents change as user trades)
        self.holdings: Dict[str, Holding] = {}
        # Stable order of symbols by first-ever successful acquisition (first fill)
        self.firstAcquired: List[str] = []
        # Orders recorded in placement order
        self.orders: List[Order] = []

    def getHoldingQty(self, stock: str) -> int:
        h = self.holdings.get(stock)
        return 0 if h is None else h.qty


class TradingService:
    def __init__(self):
        # Preserve first-added order for showStocks()
        self.stocks: "OrderedDict[str, int]" = OrderedDict()
        self.users: Dict[str, User] = {}
        # Track OPEN LIMIT orders per stock for instant re-eval on price updates
        self.openLimitByStock: Dict[str, List[OrderRef]] = {}

    # Create a user with an initial cash balance
    def addUser(self, userId: str, openingCash: int) -> None:
        if userId not in self.users:
            self.users[userId] = User(userId, openingCash)

    # Fetch user's current account balance
    def getAccountBalance(self, userId: str) -> int:
        u = self.users.get(userId)
        return 0 if u is None else u.cash

    # Add/update a stock price; on update, auto-fill favorable OPEN LIMIT orders
    def addStock(self, stockName: str, stockPrice: int) -> None:
        if stockName not in self.stocks:
            # first time: preserve insertion order
            self.stocks[stockName] = stockPrice
        else:
            # update existing price (order preserved)
            self.stocks[stockName] = stockPrice
        self._autoFillOpenLimits(stockName, stockPrice)

    # Show all stocks as "name price" in first-added order
    def showStocks(self) -> List[str]:
        return [f"{name} {price}" for name, price in self.stocks.items()]

    # Place an order; returns FILLED, OPEN, or REJECTED
    def submitOrder(
        self,
        userId: str,
        stockName: str,
        orderKind: str,
        action: str,
        qty: int,
        limitPrice: int,
    ) -> str:
        user = self.users.get(userId)
        currentPrice: Optional[int] = self.stocks.get(stockName)

        # If user doesn't exist, reject without recording
        if user is None:
            return Status.REJECTED.name

        # If stock unknown or invalid qty, record REJECTED for this user
        if currentPrice is None or qty <= 0:
            kind = OrderKind[orderKind]
            act = Action[action]
            ord_ = Order(stockName, act, qty, kind, Status.REJECTED, limitPrice)
            user.orders.append(ord_)
            return ord_.status.name

        kind = OrderKind[orderKind]
        act = Action[action]
        priceAtSubmit = currentPrice

        if kind == OrderKind.MARKET:
            return self._handleMarket(user, stockName, act, qty, priceAtSubmit)
        else:  # LIMIT
            return self._handleLimit(user, stockName, act, qty, limitPrice, priceAtSubmit)

    # Show holdings for a user as "stockName count" ordered by first-ever successful fill time
    def viewStockHoldings(self, userId: str) -> List[str]:
        user = self.users.get(userId)
        if user is None:
            return []
        out: List[str] = []
        for sym in user.firstAcquired:  # stable first-filled order
            h = user.holdings.get(sym)
            if h and h.qty > 0:
                out.append(f"{sym} {h.qty}")
        return out

    # Show order history as "stockName action quantity orderKind status"
    def viewOrders(self, userId: str) -> List[str]:
        user = self.users.get(userId)
        if user is None:
            return []
        return [
            f"{o.stockName} {o.action.name} {o.qty} {o.kind.name} {o.status.name}"
            for o in user.orders
        ]

    # --- Helpers ------------------------------------------------------------

    def _handleMarket(self, user: User, stock: str, act: Action, qty: int, priceAtSubmit: int) -> str:
        if act == Action.BUY:
            cost = qty * priceAtSubmit
            if user.cash >= cost:
                self._applyBuy(user, stock, qty, priceAtSubmit)
                status = Status.FILLED
            else:
                status = Status.REJECTED
        else:  # SELL
            held = user.getHoldingQty(stock)
            if held >= qty:
                self._applySell(user, stock, qty, priceAtSubmit)
                status = Status.FILLED
            else:
                status = Status.REJECTED

        ord_ = Order(stock, act, qty, OrderKind.MARKET, status, 0)
        user.orders.append(ord_)
        return status.name

    def _handleLimit(
        self, user: User, stock: str, act: Action, qty: int, limitPrice: int, priceAtSubmit: int
    ) -> str:
        if act == Action.BUY:
            # Must have enough cash at CURRENT price, even if not favorable
            costAtCurrent = qty * priceAtSubmit
            if user.cash < costAtCurrent:
                status = Status.REJECTED
            elif priceAtSubmit <= limitPrice:
                self._applyBuy(user, stock, qty, priceAtSubmit)
                status = Status.FILLED
            else:
                status = Status.OPEN
        else:  # SELL
            held = user.getHoldingQty(stock)
            if held < qty:
                status = Status.REJECTED
            elif priceAtSubmit >= limitPrice:
                self._applySell(user, stock, qty, priceAtSubmit)
                status = Status.FILLED
            else:
                status = Status.OPEN

        ord_ = Order(stock, act, qty, OrderKind.LIMIT, status, limitPrice)
        user.orders.append(ord_)
        if status == Status.OPEN:
            self.openLimitByStock.setdefault(stock, []).append(OrderRef(user.userId, ord_))
        return status.name

    # Auto-fill OPEN LIMIT orders for this stock if the new price is favorable
    def _autoFillOpenLimits(self, stockName: str, newPrice: int) -> None:
        refs = self.openLimitByStock.get(stockName)
        if not refs:
            return

        i = 0
        while i < len(refs):
            ref = refs[i]
            user = self.users.get(ref.userId)
            ord_ = ref.order

            # Skip/clean invalid references
            if user is None or ord_.status != Status.OPEN or ord_.kind != OrderKind.LIMIT:
                refs.pop(i)
                continue

            favorable = (ord_.action == Action.BUY and newPrice <= ord_.limitPrice) or \
                        (ord_.action == Action.SELL and newPrice >= ord_.limitPrice)

            if not favorable:
                i += 1
                continue

            if ord_.action == Action.BUY:
                cost = ord_.qty * newPrice
                if user.cash >= cost:
                    self._applyBuy(user, stockName, ord_.qty, newPrice)
                    ord_.status = Status.FILLED
                    refs.pop(i)
                    continue
                # else: keep OPEN
                i += 1
            else:  # SELL
                held = user.getHoldingQty(stockName)
                if held >= ord_.qty:
                    self._applySell(user, stockName, ord_.qty, newPrice)
                    ord_.status = Status.FILLED
                    refs.pop(i)
                    continue
                # else: keep OPEN
                i += 1

        if not refs:
            self.openLimitByStock.pop(stockName, None)

    def _applyBuy(self, user: User, stock: str, qty: int, price: int) -> None:
        cost = qty * price
        user.cash -= cost

        h = user.holdings.get(stock)
        if h is None:
            h = Holding()
            user.holdings[stock] = h

        # Record first-ever acquisition order (stable, even if later sold to zero)
        if stock not in user.firstAcquired:
            user.firstAcquired.append(stock)

        totalCostBefore = h.avgPrice * h.qty
        newQty = h.qty + qty
        newAvg = (totalCostBefore + cost) // newQty  # integer average like Java
        h.qty = newQty
        h.avgPrice = newAvg

    def _applySell(self, user: User, stock: str, qty: int, price: int) -> None:
        proceeds = qty * price
        user.cash += proceeds
        h = user.holdings.get(stock)
        if h is None or h.qty < qty:
            return  # should not happen after validation
        h.qty -= qty
        if h.qty == 0:
            # remove empty holding but DO NOT remove from firstAcquired
            user.holdings.pop(stock, None)
        # avgPrice for remaining shares stays as-is
